### ISSUE TITLE
os: regression in os.Expand treatment of '${}'

### ISSUE THREAD

**Issue by iand** [golang/go#26135]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?
tip

### Does this issue reproduce with the latest release?
yes

### What operating system and processor architecture are you using (`go env`)?
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/iand/.cache/go-build"
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/iand"
GORACE=""
GOROOT="/opt/go"
GOTMPDIR=""
GOTOOLDIR="/opt/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build689803627=/tmp/go-build -gno-record-gcc-switches"

### What did you do?

```
package main

import (
	"fmt"
	"os"
)

func main() {
	fmt.Printf("%q", substituteEnv("${}"))
}

func substituteEnv(s string) string {
	return os.Expand(s, func(key string) string {
		if val, ok := os.LookupEnv(key); ok {
			return val
		}
		return key
	})
}

```

### What did you expect to see?
An empty string printed.  The Go 1.10 behaviour is that os.Expand returns an empty string.

Can be see on playground which is currently running Go 1.10.2

### What did you see instead?

The $ is printed without the braces.

The CL that introduced the new behaviour: https://github.com/golang/go/commit/7bf631e1fc55839a6bff404a91597326e1addc10 derived from https://github.com/golang/go/issues/24345

**Comment by agnivade** [comment]
Well, that was intentional. Not sure if something can be done about it. /cc @ianlancetaylor

**Comment by iand** [comment]
I think I'd expect it to either output `${}` or an empty string. `$` on it's own seems wrong.

**Comment by mvdan** [comment]
None of the added test cases in that CL test for `${}`, nor do they make any characters disappear, so I think this was a mistake. Will look into a fix.

**Comment by agnivade** [comment]
Thanks @mvdan. I was afraid we would be adding checks for more edge cases. But I guess it's better not to break existing behavior.

**Comment by mvdan** [comment]
@agnivade indeed. I'm more concerned about consistency than existing behavior, but both are important. See my comment on the CL for an alternative approach that doesn't keep existing behavior.
